File: backend/app/schemas/service.py
"""Service schemas."""

import json
from datetime import date, datetime

from pydantic import BaseModel, Field, field_validator

# AUT-1275: "Oil Change" service type is merged into "Scheduled Service" across
# the entire stack. Legacy values are normalised on ingress so stored records
# stay consistent. The _LEGACY_OIL_TYPES set covers every variant that may
# exist in the DB (seed data uses "oil", old clients used "oil_change").
_LEGACY_OIL_TYPES = frozenset({"oil", "oil_change"})


class ServiceItemIn(BaseModel):
    part_id: str | None = None
    name: str = Field(min_length=1, max_length=255)
    quantity: int = 1
    unit_cost: float = 0.0
    kind: str = "item"  # part/labour/item
    part_no: str | None = None
    labour_hours: float | None = None
    labour_rate: float | None = None


class ServiceItemOut(BaseModel):
    id: str
    name: str
    quantity: int
    unit_cost: float
    kind: str
    part_no: str | None
    part_id: str | None
    labour_hours: float | None
    labour_rate: float | None

    model_config = {"from_attributes": True}


class ServiceCreate(BaseModel):
    service_date: date
    odometer_km: int
    service_type: str = "scheduled"
    description: str | None = None
    workshop: str | None = None
    cost: float = 0.0
    currency: str = "AUD"
    notes: str | None = None
    status: str = Field(default="completed", pattern="^(scheduled|completed)$")
    steps: list[str] = []
    items: list[ServiceItemIn] = []

    @field_validator("service_type")
    @classmethod
    def _merge_oil_change(cls, v: str) -> str:
        return "scheduled" if v in _LEGACY_OIL_TYPES else v


class ServiceUpdate(BaseModel):
    service_date: date | None = None
    odometer_km: int | None = None
    service_type: str | None = None
    description: str | None = None
    workshop: str | None = None
    cost: float | None = None
    currency: str | None = None
    notes: str | None = None
    status: str | None = Field(default=None, pattern="^(scheduled|completed)$")
    completed_date: date | None = None
    steps: list[str] | None = None
    items: list[ServiceItemIn] | None = None

    @field_validator("service_type")
    @classmethod
    def _merge_oil_change(cls, v: str | None) -> str | None:
        return "scheduled" if v in _LEGACY_OIL_TYPES else v


class ServiceOut(BaseModel):
    id: str
    vehicle_id: str
    service_date: date
    odometer_km: int
    service_type: str
    description: str | None
    workshop: str | None
    cost: float
    currency: str
    notes: str | None
    ai_prediction: str | None
    next_due_km: int | None
    next_due_date: date | None
    status: str
    completed_date: date | None
    steps: list[str] = []
    items: list[ServiceItemOut] = []
    created_at: datetime

    @field_validator("steps", mode="before")
    @classmethod
    def _parse_steps(cls, v):
        if isinstance(v, str):
            try:
                return json.loads(v)
            except Exception:
                return []
        return v or []

    model_config = {"from_attributes": True}


class ServicePredictionRequest(BaseModel):
    make: str
    model: str
    year: int
    odometer_km: int
    last_service_km: int | None = None
    last_service_days_ago: int | None = None
    service_type: str = "scheduled"

    @field_validator("service_type")
    @classmethod
    def _merge_oil_change(cls, v: str) -> str:
        return "scheduled" if v in _LEGACY_OIL_TYPES else v


class ServicePredictionResponse(BaseModel):
    service_type: str
    interval_km: int
    interval_months: int
    due_in_km: int
    due_in_days: int
    next_due_km: int
    next_due_date: date
    confidence: float
    reason: str

File: frontend/lib/core/models/service.dart
part of models;

class ServiceItem {
  final String id;
  final String name;
  final int quantity;
  final double unitCost;
  final String kind;
  final String? partNo;
  final String? partId;

  const ServiceItem({
    required this.id,
    required this.name,
    this.quantity = 1,
    this.unitCost = 0,
    this.kind = 'item',
    this.partNo,
    this.partId,
  });

  double get total => quantity * unitCost;

  factory ServiceItem.fromJson(Map<String, dynamic> j) => ServiceItem(
        id: j['id'] as String,
        name: j['name'] as String,
        quantity: (j['quantity'] as num?)?.toInt() ?? 1,
        unitCost: (j['unit_cost'] as num?)?.toDouble() ?? 0,
        kind: j['kind'] as String? ?? 'item',
        partNo: j['part_no'] as String?,
        partId: j['part_id'] as String?,
      );
}

class ServiceRecord {
  final String id;
  final String serviceDate;
  final int odometerKm;
  final String serviceType;
  final String? description, workshop, notes;
  final double cost;
  final int? nextDueKm;
  final String? nextDueDate;
  final String status; // scheduled/completed
  final String? completedDate;
  final List<String> steps;
  final List<ServiceItem> items;

  const ServiceRecord({
    required this.id,
    required this.serviceDate,
    required this.odometerKm,
    required this.serviceType,
    this.description,
    this.workshop,
    this.notes,
    this.cost = 0,
    this.nextDueKm,
    this.nextDueDate,
    this.status = 'completed',
    this.completedDate,
    this.steps = const [],
    this.items = const [],
  });

  bool get isScheduled => status == 'scheduled';

  factory ServiceRecord.fromJson(Map<String, dynamic> j) => ServiceRecord(
        id: j['id'] as String,
        serviceDate: j['service_date'] as String,
        odometerKm: j['odometer_km'] as int,
        serviceType: j['service_type'] as String,
        description: j['description'] as String?,
        workshop: j['workshop'] as String?,
        notes: j['notes'] as String?,
        cost: (j['cost'] as num?)?.toDouble() ?? 0,
        nextDueKm: j['next_due_km'] as int?,
        nextDueDate: j['next_due_date'] as String?,
        status: j['status'] as String? ?? 'completed',
        completedDate: j['completed_date'] as String?,
        steps: ((j['steps'] as List?) ?? []).map((e) => e.toString()).toList(),
        items: ((j['items'] as List?) ?? [])
            .map((e) => ServiceItem.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class ServicePrediction {
  final String serviceType;
  final int intervalKm, dueInKm, nextDueKm;
  final String nextDueDate;
  final double confidence;
  final String reason;

  const ServicePrediction({
    required this.serviceType,
    required this.intervalKm,
    required this.dueInKm,
    required this.nextDueDate,
    required this.confidence,
    required this.reason,
    this.nextDueKm = 0,
  });

  factory ServicePrediction.fromJson(Map<String, dynamic> j) =>
      ServicePrediction(
        serviceType: j['service_type'] as String,
        intervalKm: (j['interval_km'] as num).toInt(),
        dueInKm: (j['due_in_km'] as num).toInt(),
        nextDueDate: j['next_due_date'] as String,
        confidence: (j['confidence'] as num).toDouble(),
        reason: j['reason'] as String,
        nextDueKm: (j['next_due_km'] as num?)?.toInt() ?? 0,
      );
}
